### PR TITLE
[SPARK-32217] Plumb whether a worker would also be decommissioned along with executor

### PR DESCRIPTION
**Pull request by agrawaldevesh**
### What changes were proposed in this pull request?

This PR is a giant plumbing PR that plumbs an `ExecutorDecommissionInfo` along
with the DecommissionExecutor message.

### Why are the changes needed?

The primary motivation is to know whether a decommissioned executor
would also be loosing shuffle files -- and thus it is important to know
whether the host would also be decommissioned. 

In the absence of this PR, the existing code assumes that decommissioning an executor does not loose the whole host with it, and thus does not clear the shuffle state if external shuffle service is enabled. While this may hold in some cases (like K8s decommissioning an executor pod, or YARN container preemption), it does not hold in others like when the cluster is managed by a Standalone Scheduler (Master). This is similar to the existing `workerLost` field in the `ExecutorProcessLost` message. 

In the future, this `ExecutorDecommissionInfo` can be embellished for
knowing how long the executor has to live for scenarios like Cloud spot
kills (or Yarn preemption) and the like.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tweaked an existing unit test in `AppClientSuite`

### CODE REVIEW

**Comment by agrawaldevesh** [comment]
@holdenk, @jiangxb1987 @cloud-fan @Ngone51 -- This PR is ready for your review please. Thanks !

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:422]: nit: 4 indent.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:590]: ditto.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:139]: This is not used at all?

File: core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala:27]: Maybe, `workerDecommissioned`?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala:27]: We probably need a base class for decom or just extends `ExecutorLossReason`... Then, we could use the base class or `ExecutorLossReason` in `TaskScheduler.executorDecommission`. Thereform, `TaskScheduler.executorDecommission` could support other cluster managers as well.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:139]: ok. I see you used it in #29014. But probably it's better to remove it here. You can rebase #29014 later if this PR gets merged.

Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:139]: I was debating about this. I think I am now convinced of moving it to #29014 :-).

File: core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala:27]: DecommissionInfo is just an `info` struct. It's not the loss reason itself, so it shouldn't overload ExecutorLossReason. 

I didn't get the part about TaskScheduler.executorDecommission supporting other cluster managers: TaskScheduler is about task scheduling not coarse grained executor scheduling. I think it should be totally unaware of whether the decommission signal came via Yarn or K8s or whatever other cluster management strategy we have.

**Comment by agrawaldevesh** [comment]
Retest this please.

**Comment by agrawaldevesh** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala]: But now, it looks like `DecommissionInfo` is specific to Standalone because of the terminology “worker”. 

I agree that `TaskScheduler` should be unaware of where the decom signal comes. That's why I asked for the base decom class. However, it's still ok if `DecommissionInfo` could be applied to other clusters as well.

**Comment by Ngone51** [comment]
ok to test

**Comment by cloud-fan** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:915]: nit: code style should be
```
override def ...(
    para1: T, para2: T): ... 
```

4 space indentation for the parameter list.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:718]: ditto: indentation

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:93]: ditto

File: core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala:27]: so this PR is just a refactor and doesn't actually use the `isWorkerDecommissioned` flag?

File: core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala]: Fair enough. I can see why the "Worker" nomenclature is confusing. Although `ExecutorProcessLost` has a `workerLost` field that can (in theory) be set by any scheduler backend. I will rename/reword to make this independance clearer.

File: core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/DecommissionInfo.scala:27]: Yeah. Its actually used in https://github.com/apache/spark/pull/29014 (which builds atop this one). I pulled the usage part to that PR since it was more closely coupled there.

File: core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala:104]: nit: don't leave an empty implementation here.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:195]: do we want to also include the decommissionInfo in the error msg?

File: core/src/main/scala/org/apache/spark/deploy/client/StandaloneAppClient.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/client/StandaloneAppClient.scala:186]: how is the flag `isHostDecommissioned` actually used?

Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/deploy/client/StandaloneAppClient.scala:186]: oh I see https://github.com/apache/spark/pull/29032#discussion_r455401121

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala:104]: yea we should make sure it's implemented

Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala:104]: Okay. Was just trying to avoid changing a bunch of mock/test implementations of this interface. Will revert back to having this be purely abstract.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by agrawaldevesh** [comment]
@holdenk ... It would be great for you to review this PR. This PR stems from your suggestion of plumbing a "isWorkerLost" to the executor decommission message, as you suggested in https://github.com/apache/spark/pull/29014#issuecomment-654420917. It does not introduce any other semantic changes. 

I am keeping this as a separate PR instead of folding it into https://github.com/apache/spark/pull/29014, to keep the latter easy to revert. This PR (#29032) shouldn't be introducing any semantic changes and thus should be less risky.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala:874]: Why did this need to change to true?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:170]: Is there a reason we broke this on two lines (here and elsewhere)? I don't have strong feelings just if theres been a change to the style want to make sure I'm in the loop.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:174]: minor style nit: I like to pass by name when I'm passing a bare boolean, I think it makes the code easier to read.

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:25]: How about "even if the external shuffle service" for clarity?

**Review by holdenk — COMMENTED** [review]
Thanks for working on this. A few minor questions and nits for cleanup but I think it is really close :) Thanks for factoring this out into a seperate PR for simpler reviewing as well, that's always appreciated :)

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/deploy/master/Master.scala:874]: The short answer is that I sort of hacked it here :-)

So on line 186 of StandaloneAppClient.scala in this PR: I actually plumb the workerLost (= true) here into ExecutorDecommissionInfo.isHostDecommissioned. 

The proper thing to do here would be to not communicate executor decommissioning via ExecutorUpdated in the standalone codepath. But instead have a separate ExecutorDecommissioned message that can take the ExecutorDecommissionInfo. The ExecutorDecommissionInfo should be prepared by the cluster manager directly, since it has the most context about the decom is happening (ie exactly what pieces would be going away). 

This is just a slightly bigger change and maybe I should just do it here. Let me try it.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:174]: Good idea.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:170]: It's not broken into two lines for style. I wanted to reuse the "msg" down below. Same deal elsewhere in this PR.

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:25]: Changed it to "in case" which felt more accurate: Barring a few deployments, external shuffle service is host local: So if a host is decommissioned then its shuffle service will invariably go and thus so would its shuffle.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/deploy/master/Master.scala:874]: On second thoughts, this is a bigger change: I will just explain why workerLost = true better in the code. 

The main problem is that the Deploy messages are b/w the Master and the Driver, while the ExecutorDecommissionInfo are b/w the Driver and the Executors. In any case, each cluster manager would need to have its own way of somehow communicating the decommissioning metadata to the driver, and this workerLost could be the Master's internal way.

The alternatives would be to either introduce another `ExecutorDecommissioned(id, MasterToDriverExecutorDecommissionInfo)` struct, or enhance `ExecutorUpdated.workerLost` to be a full Struct/Any that can take the MasterToDriverExecutorDecommissionInfo. 

This all seems like a bit too much machinery for now, so I will just comment it inline in the code. 

Thank you for raising this question :-)

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:25]: Actually, sorry ... "even if" is indeed okay here :-)

**Comment by agrawaldevesh** [comment]
@holdenk, this is ready again for your review. Thanks !

**Comment by holdenk** [comment]
The pyspark packaging tests have generally been flaky so i think we don't have to worry about them given that the GH actions all passed. Let me do another pass through reveiw, although it looks like it's already got one approval :)

**Comment by holdenk** [comment]
Thanks for the updates @agrawaldevesh. This looks good to me. I don't see any pending conversations that are ongoing, so if no one has any concerns that they aren't ok carrying through to the follow on PR that you've already got open, I plan to merge this tomorrow.

**Comment by Ngone51** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:28]: +1 for the renaming.

**Comment by Ngone51** [comment]
LGTM

**Comment by agrawaldevesh** [comment]
retest this please

**Comment by agrawaldevesh** [comment]
@holdenk... this is ready to go in. Thanks !

**Comment by holdenk** [comment]
Cool I’ll try merge this after dinner, thanks everyone for working on and reviewing this 👍

**Comment by cloud-fan** [comment]
I have a general question (not related to this PR): We have deprecated the usage of "multi-workers on one host", so host and worker should be the same thing. However, in the codebase sometimes we use `host` and sometimes `worker`, do we have a general rule for the naming?

This PR LGTM, I just ask this question for curiosity.

**Comment by holdenk** [comment]
Merged, thanks everyone.

I don't know if we have a good principle around naming host v. worker with the deprecation.

**Comment by agrawaldevesh** [comment]
> Merged, thanks everyone.
> 
> I don't know if we have a good principle around naming host v. worker with the deprecation.

Thanks @holdenk for championing and shepherding this change !!

@cloud-fan, I agree that this host <=> worker naming is confusing. It was especially confusing for a newcomer to me: `ExecutorProcessLost` has a field `workerLost` (renamed from the erstwhile `slaveLost`). I had to pick b/w `Worker` and `Host` in `ExecutorDecommissionInfo.isHostDecommissioned` and I went with `Worker` initially but later changed `Host`. 

But in my mind, they are both the same: Basically there is one external shuffle service (if any) per worker (or host). The word `worker` might draw some unintended connotation to the Standalone mode so I refrained from using it. But I think its usage is quite pervasive already.

It would be nice to have consistent terminology around `worker` vs `host` vs `node` :-P. In the absence of any, I went with `host` here.

**Comment by cloud-fan** [comment]
I agree `host` is better as it's general to any deployment model. I think we should only use `worker` when it refers to the standalone worker. Maybe we can start following this principle from now on.

**Pull request closed without merging**
